s3/s3.cc

```diff
@@ -0,0 +1,625 @@
+        return;
+    }
+    auto ptr = std::move(_ptr);
+    log.debug("conn {}: not closed, closing in background", fmt::ptr(&*ptr));
+    (void)ptr->close()
+        .then([ptr] {
+            log.debug("conn {}: closed", fmt::ptr(&*ptr));
+        }).handle_exception([ptr](auto ep) {
+            log.error("conn {}: failed to close: {}", fmt::ptr(&*ptr), ep);
+        });
+}
+
+class client_impl : public client, public enable_shared_from_this<client_impl> {
+    connection_factory_ptr _connection_factory;
+    std::map<sstring, sstring> _headers;
+protected:
+    future<connection_ptr> new_connection() {
+        return _connection_factory->connect();
+    }
+
+    template <typename Func, typename FuncResult = std::invoke_result_t<Func, connection_ptr>>
+    futurize_t<FuncResult>
+    do_with_connection(Func func) {
+        return new_connection().then([func = std::move(func), this] (connection_ptr con) mutable {
+            return futurize_invoke(std::move(func), con)
+                    .then_wrapped([con, this] (auto&& f) mutable {
+                        if (!f.failed()) {
+                            _connection_factory->take_back(con);
+                            return std::move(f);
+                        } else {
+                            return con->close()
+                                .handle_exception([con] (auto ep) mutable {
+                                    log.error("conn {}: failed to close: {}", fmt::ptr(&*con), ep);
+                                }).then([con, f = std::move(f)] () mutable {
+                                    return std::move(f);
+                                });
+                        }
+                    });
+        });
+    }
+private:
+    class s3_upload_data_sink : public data_sink_impl {
+        // "Each part must be at least 5 MB in size, except the last part."
+        // https://docs.aws.amazon.com/AmazonS3/latest/API/API_UploadPart.html
+        static constexpr size_t minimum_part_size = 5*1024*1024;
+
+        struct part {
+            sstring etag;
+        };
+
+        client_ptr _client;
+        client_impl& _client_impl;
+        const sstring _path;
+        sstring _upload_id; // non-empty when upload started.
+        uint64_t _part_number = 0;
+        circular_buffer<temporary_buffer<char>> _current_part;
+        seastar::gate _uploads;
+        utils::chunked_vector<part> _parts;
+        utils::phased_barrier _uploads_barrier;
+        semaphore _flush_sem{2}; // allow for two concurrent uploads
+        bool _upload_done = false; // true if the upload was completed or aborted on the server side
+    private:
+        size_t current_part_size() const {
+            size_t res = 0;
+            for (auto&& buf : _current_part) {
+                res += buf.size();
+            }
+            return res;
+        }
+
+        future<> maybe_flush() {
+            if (current_part_size() >= minimum_part_size) {
```

Comment:
Why break parts at the minimum boundary? Is there guidance about how to split?

Too many splits can reduce performance due to tcp slow start.